A few changes...

ShadowNinja opened the pull request:
Send mode options to the OnModeChange hook. (For modes that use options like +bieqfovhj)
Add OnSend hook. (Logging)
Interpret format codes literally if there aren't any formating arguments.  (Don't have to escape messages to send())
Add PreRegister hook. (Allows for CAP commands and SASL)
Track user access changes and request multi-prefix. (This also changes the format of user.access)

ShadowNinja committed:
Add PreRegister hook

This is useful for CAP commands and SASL.

ShadowNinja committed:
Interpret format codes passed to meta:send() literally if there are no format arguments

ShadowNinja committed:
Add OnSend hook

ShadowNinja committed:
Send mode options to the OnModeChange hook

JakobOvrum commented:
Oops, sorry for leaving this hanging!

Indeed there is a bug with `send` and unintentional formatting options, but using straight-forward concatenation produces a lot of unnecessary garbage strings (as in _garbage collector_) when there are more than two strings being concatenated. `format` provides both optimal memory handling and convenience.

An alternative that also fixes the extraneous garbage is to use `table.concat` when there are more than two strings involved (e.g. `table.concat{"USER ", self.username, " 0 * :", self.realname}`). The pay-off here scales linearly with the number of strings being concatenated.

In the memory department, `format` has an additional upper hand over `table.concat`: it doesn't need a table allocated for the input.

As this is library code it's nice to keep unnecessary stress on the garbage collector at a minimum, as you can't predict the requirements of users.

I think the best solution would be to keep the formatting functionality but make it not do any formatting if there are no formatting arguments:

```
if select("#", ...) > 0 then
    msg = msg:format(...)
end
```

What do you think?

The other changes look good!

ShadowNinja commented:
Done. I used #{...} instead of select() because I didn't know how to get it. However this makes a table, which you probably don't want.

JakobOvrum commented:
What do you mean?

`select` is part of the Lua standard library, both 5.1 and 5.2 ([select's entry in the manual](http://www.lua.org/manual/5.1/manual.html#pdf-select)).

JakobOvrum commented on doc/irc.luadoc:
```diff
@@ -125,7 +125,9 @@ function irc:shutdown()
 
 --- List of hooks you can use with irc:hook. The parameter list describes the parameters passed to the callback function.
 -- <ul>
+-- <li><code>PreRegister(connection)</code>Usefull for CAP commands and SASL.</li>
```
Should be `Useful`

ShadowNinja committed:
Track user access changing and enable multi-prefix

This also changes the format of user.access.

ShadowNinja commented:
It seems the module command takes that all out of scope. I wasn't copying it locally at the top. It is fixed now. I have also added access changing following. In doing so I have changed the way that access is stored, but that shouldn't be a big issue.

JakobOvrum commented:
Looks good, thanks a lot!
